library/core/src/slice/sort/select.rs

```diff
@@ -298,3 +328,92 @@ fn median_idx<T, F: FnMut(&T, &T) -> bool>(
     }
     b
 }
+
+// It's possible to re-use the insertion sort in the smallsort module, but with optimize_for_size it
+// would clutter that module with cfg statements and make it generally harder to read and develop.
+// So to decouple things and simplify it, we use a an even smaller bubble sort.
+#[cfg(feature = "optimize_for_size")]
+fn bubble_sort<T, F: FnMut(&T, &T) -> bool>(v: &mut [T], is_less: &mut F) {
+    let mut n = v.len();
+    let mut did_swap = true;
+
+    while did_swap && n > 1 {
+        did_swap = false;
+        for i in 1..n {
+            // SAFETY: The loop construction implies that `i` and `i - 1` will always be in-bounds.
+            unsafe {
+                if is_less(v.get_unchecked(i), v.get_unchecked(i - 1)) {
+                    v.swap_unchecked(i - 1, i);
+                    did_swap = true;
+                }
+            }
+        }
+        n -= 1;
+    }
+}
+
+#[cfg(feature = "optimize_for_size")]
+fn partition_size_opt<T, F>(v: &mut [T], pivot: usize, is_less: &mut F) -> usize
```

Comment:
This likely actually applies a few other places. For example, the selection between `insertion_sort_shift_left` and `bubble_sort`, among others.

These actual implementation functions could be kept relatively private, then introduce new functions `sort_impl`/`sort_unstable_impl` that more public and just selects between the two with `cfg`. Tests should test `sort_impl` rather than the individual functions, and then anything that needs it can just call `sort_impl` without duplicating the config.

This way we more or less know that the implementations are the same (they pass the same tests), and anything that calls it doesn't need to worry about `cfg` or what the actual implementation is (not that I think these get called in multiple places, but they could).

Reply:
As said above, I think it's a bad idea to test these private implementation details, and so they should stay private. I prefer different names because it makes it clear that logic diverges here, Rust cares a lot about explicitness and I'd prefer to make it very clear to future readers that logic diverges depending on `optimize_for_size` where it does.

That said, thinking about it a bit more, I think I found a way to reduce code duplication in this specific place.